File: lib/skullstrip/skullstrip_epi.m
function mask_array = skullstrip_epi(input, pathSPM, roi_size, scale, nerode, ndilate, savemask)
% Skullstrip input volume by defining an intensity threshold from the inner 
% of the brain volume. From a defined mid-point, a brain mask is grown 
% inside the brain. A binary filling holes algorithm is applied. To reduce 
% remaining skull within the brain mask, the mask is eroded and dilated 
% several times.    
% Inputs:
%   *input: input file.
%   *pathSPM: path to spm toolbox.
%   *roi_size: size of cubic roi for image intensity threshold
%   *scale: scale image intensity threshold
%   *nerode: number of eroding iterations
%   *ndilate: number of dilating iterations
%   *savemask: save mask time series (boolean).
%   *cleanup: delete intermediate files after running.
%    
% created by Daniel Haenelt
% Date created: 06-11-2018             
% Last modified: 25-02-2019

% default parameters
if ~exist('roi_size','var')
    roi_size = 10;
end

if ~exist('scale','var')
    scale = 0.7;
end

if ~exist('nerode','var')
    nerode = 3;
end

if ~exist('ndilate','var')
    ndilate = 1;
end

if ~exist('savemask','var')
    savemask = true;
end

% add spm toolbox
addpath(pathSPM);

% prepare path and filename
[path, file, ext] = fileparts(input);

% load data
data_img = spm_vol(input);
data_array = spm_read_vols(data_img);

% calculate mean intensity
data_mean = mean(data_array(:));

% get point within the brain
[x, y, z] = ind2sub(size(data_array),find(data_array > data_mean));
x_mean = round((max(x)+min(x))/2);
y_mean = round((max(y)+min(y))/2);
z_mean = round((max(z)+min(z))/2);

% initialize mask
mask_array = zeros(size(data_array));
mask_temp_array = zeros(size(data_array));
mask_array(x_mean,y_mean,z_mean) = 1;
mask_temp_array(x_mean,y_mean,z_mean) = 1;

% compute threshold
roi = data_array(...
    ceil(x_mean-roi_size/2):floor(x_mean+(roi_size-1)/2),...
    ceil(y_mean-roi_size/2):floor(y_mean+(roi_size-1)/2),...
    ceil(z_mean-roi_size/2):floor(z_mean+(roi_size-1)/2)...
    );
roi_mean = mean(roi(:));

% grow mask
while true
    
    % get ROI coordinates
    [x, y, z] = ind2sub(size(mask_array),find(mask_array > 0));
    coords = [x y z];
    
    % edge
    coords = coords(coords(:,1) > 1,:);
    coords = coords(coords(:,2) > 1,:);
    coords = coords(coords(:,3) > 1,:);
    coords = coords(coords(:,1) < size(data_array,1)-1,:);
    coords = coords(coords(:,2) < size(data_array,2)-1,:);
    coords = coords(coords(:,3) < size(data_array,3)-1,:);
        
    % calculate nearest neighbors
    coords1 = [coords(:,1)-1, coords(:,2), coords(:,3)];
    coords2 = [coords(:,1), coords(:,2)-1, coords(:,3)];
    coords3 = [coords(:,1), coords(:,2), coords(:,3)-1];
    coords4 = [coords(:,1)+1, coords(:,2), coords(:,3)];
    coords5 = [coords(:,1), coords(:,2)+1, coords(:,3)];
    coords6 = [coords(:,1), coords(:,2), coords(:,3)+1];
    
    % merge new coordinates
    coords = [coords ; coords1 ; coords2 ; coords3 ; coords4 ; coords5 ; coords6];
    
    % convert to indices
    inds = sub2ind(size(mask_array),coords(:,1),coords(:,2),coords(:,3));
    
    % get temporary mask from coordinates
    mask_temp_array(inds) = 1;
       
    % delete all old mask elements
    mask_temp_array(mask_array == 1) = 0;
    mask_temp_array(data_array < scale*roi_mean) = 0;
       
    % reinitialize mask temp
    mask_array = mask_array + mask_temp_array;
       
    % check break condition
    if length(find(mask_temp_array > 0)) == 0
        break
    end
    
end

% flood filling brain mask
mask_array = imfill(mask_array,'holes');

% erode mask
for i = 1:nerode
    mask_array = imerode(mask_array,ones(3));
end

% dilate mask
for i = 1:ndilate
    mask_array = imdilate(mask_array, ones(3));
end
    
% write maskedimage
data_array = data_array .* mask_array;
data_img.fname = fullfile(path, ['p' file ext]);
spm_write_vol(data_img, data_array);

% write mask
if savemask
    data_img.fname = fullfile(path, ['mask' file ext]);
    spm_write_vol(data_img, mask_array);
end

File: fmri_tools/skullstrip/skullstrip_epi.py
# -*- coding: utf-8 -*-

# python standard library inputs
import os

# external inputs
import numpy as np
import nibabel as nb
from nipype.interfaces.fsl import ErodeImage, DilateImage
from scipy.ndimage.morphology import binary_fill_holes


def skullstrip_epi(file_in, roi_size=5, scale=0.75, nerode=2, ndilate=1,
                   savemask=False, cleanup=True):
    """Skullstrip EPI.

    Skullstrip input volume by defining an intensity threshold from the inner of 
    the brain volume. From a defined mid-point, a brain mask is grown inside the 
    brain. A binary filling holes algorithm is applied. To reduce remaining 
    skull within the brain mask, the mask is eroded and dilated several times.        

    Parameters
    ----------
    file_in : str
        Input file.
    roi_size : int, optional
        Size of cubic roi for image intensity threshold. The default is 5.
    scale : float, optional
        Scale image intensity threshold. The default is 0.75.
    nerode : int, optional
        Number of eroding iterations. The default is 2.
    ndilate : int, optional
        Number of dilating iterations. The default is 1.
    savemask : bool, optional
        Save mask time series. The default is False.
    cleanup : bool, optional
        Delete intermediate files after running. The default is True.

    Returns
    -------
    None.
    
    """
    
    # prepare path and filename
    path = os.path.split(file_in)[0]
    file = os.path.split(file_in)[1]

    # load data
    data_img = nb.load(os.path.join(path, file))
    data_array = data_img.get_data()
    
    # calculate mean intensity
    data_mean = data_array.mean()

    # get point within the brain
    inds = np.transpose(np.nonzero(data_array > data_mean))    
    x_mean = np.uint8(np.round((np.max(inds[:, 0])+np.min(inds[:, 0]))/2))
    y_mean = np.uint8(np.round((np.max(inds[:, 1])+np.min(inds[:, 1]))/2))
    z_mean = np.uint8(np.round((np.max(inds[:, 2])+np.min(inds[:, 2]))/2))
    
    # initialise mask
    mask_array = np.zeros_like(data_array)
    mask_temp_array = np.zeros_like(data_array)
    mask_array[x_mean, y_mean, z_mean] = 1
    mask_temp_array[x_mean, y_mean, z_mean] = 1
        
    # compute threshold
    roi = data_array[np.uint8(np.round(x_mean-roi_size/2)):np.uint8(np.round(x_mean+roi_size-1/2)),
                     np.uint8(np.round(y_mean-roi_size/2)):np.uint8(np.round(y_mean+roi_size-1/2)),
                     np.uint8(np.round(z_mean-roi_size/2)):np.uint8(np.round(z_mean+roi_size-1/2))]
    roi_mean = roi.mean()

    # grow mask
    while True:
            
        coords = np.transpose(np.nonzero(mask_array > 0))
    
        coords = coords[coords[:, 0] > 0, :]
        coords = coords[coords[:, 1] > 0, :]
        coords = coords[coords[:, 2] > 0, :]
        coords = coords[coords[:, 0] < np.size(data_array, 0)-1]
        coords = coords[coords[:, 1] < np.size(data_array, 1)-1]
        coords = coords[coords[:, 2] < np.size(data_array, 2)-1]
   
        # calculate neighbour coordinate
        mask_temp_array[coords[:, 0]-1, coords[:, 1], coords[:, 2]] = 1
        mask_temp_array[coords[:, 0], coords[:, 1]-1, coords[:, 2]] = 1
        mask_temp_array[coords[:, 0], coords[:, 1], coords[:, 2]-1] = 1
        mask_temp_array[coords[:, 0]+1, coords[:, 1], coords[:, 2]] = 1
        mask_temp_array[coords[:, 0], coords[:, 1]+1, coords[:, 2]] = 1
        mask_temp_array[coords[:, 0], coords[:, 1], coords[:, 2]+1] = 1
   
        # delete all old mask elements
        mask_temp_array = mask_temp_array - mask_array      
        mask_temp_array[data_array < scale*roi_mean] = 0
    
        # reinitialise mask_temp
        mask_array = mask_array + mask_temp_array

        # check break condition
        coords = np.transpose(np.nonzero(mask_temp_array == True))
        if len(coords) == 0:
            break

    # flood filling on brain mask
    mask_array = binary_fill_holes(mask_array, structure=np.ones((2, 2, 2)))

    # write mask (intermediate)
    newimg = nb.Nifti1Image(mask_array, data_img.affine, data_img.header)
    newimg.header['dim'][0] = 3
    nb.save(newimg, os.path.join(path, 'temp.nii'))

    # erode mask
    for i in range(nerode):
        erode = ErodeImage()
        erode.inputs.in_file = os.path.join(path, 'temp.nii')
        erode.inputs.output_type = 'NIFTI'
        erode.inputs.out_file = os.path.join(path, 'temp.nii')
        erode.run()

    # dilate mask
    for i in range(ndilate):
        dilate = DilateImage()
        dilate.inputs.in_file = os.path.join(path, 'temp.nii')
        dilate.inputs.operation = 'mean'
        dilate.inputs.output_type = 'NIFTI'
        dilate.inputs.out_file = os.path.join(path, 'temp.nii')
        dilate.run()
        
    # load final mask
    temp_img = nb.load(os.path.join(path, 'temp.nii'))
    mask_array = temp_img.get_data()
    
    # write masked image
    data_masked_array = data_array * mask_array
    output = nb.Nifti1Image(data_masked_array, data_img.affine, data_img.header)
    nb.save(output, os.path.join(path, 'p'+file))
    
    # write final output
    if savemask is True:
        newimg = nb.Nifti1Image(mask_array, data_img.affine, data_img.header)
        nb.save(newimg, os.path.join(path, 'mask_'+file))
    
    # clear output
    if cleanup is True:
        os.remove(os.path.join(path, 'temp.nii'))
